### ISSUE TITLE
Some questions about this plug-in

### ISSUE THREAD

**Issue by Phillipus** [archi-contribs/toolbox-plugin#1]
Hi Herve,

I've been taking a look at your code here in the `dev` branch and I'm a bit concerned by your approach. It seems you intend to over-ride and patch Archi's main code through Java introspection and then compile and save new patched jar files to Archi's installation. Is that correct?

Phil

**Comment by herve91** [comment]
yes, that's correct.

Unfortunately, your code contains to many private methods and variables and therefore I cannot override them. My only choice is to replace them.

For now, I'm developping two functionalities this way:
- protection: graphical interface will not alow to delete or update protected archimate concepts.
- uniqueness: graphical interface will not alow to have several archimate class of same class having the same name.

**Comment by Phillipus** [comment]
> Unfortunately, your code contains to many private methods and variables and therefore I cannot override them.

Not "unfortunately", there's a reason for this as these are not API and not meant to be over-ridden. As a matter of fact I intend to make several classes `final` or `sealed` in the main Archi code to prevent possible conflicts.

Your approach is dangerous for a number of reasons:

 - Security issues of modifying jars (in future these may be signed)
 - On Mac, the app is signed and may not run at all if any of its files are modified
 - On Windows, Archi installs in read-only directories and requires admin access to modify files
 - Over-riding Editors, Views, and UI Providers can lead to conflicts - especially if more than one plug-in does this, as the order of loading plug-ins is indeterminate

If you intend to only use this plug-in privately then OK, but I cannot endorse this approach or this plug-in and will advise Archi users not to install it as there are too many dangers. If you intend to continue with this approach then I'm sorry but we can't host this repository here at `archi-contribs` (as it implies endorsement) and you'll have to host it elsewhere.

Phil

**Comment by herve91** [comment]
You told me that you wouldn't accept a pull request on these functionalities, and I canoot use Eclipse mechanism to override only few methods without rewriting full classes.

May be you can provide guidance about a third way to achieve them without Java introspection ?

**Comment by Phillipus** [comment]
> May be you can provide guidance about a third way to achieve them without Java introspection ?

As there are currently no extension points one must implement them in the core code.  I won't be implementing this any time soon as I refuse to be pushed into implementing features that are not fully scoped out and without full impact analysis on how this affects all features of Archi, coArchi and jArchi.

**Comment by jbsarrodie** [comment]
Hi guys,

I fully agree with Phil's analysis: There's always a good reason for some methods to be private and thus outside the API accessible by a plugin. Going against this will lead to lots of troubles for users of this plugin. This means that I also agree with the conclusion that this plugin should not be part of the archi-contribs "organization" (which implicitely suggest that Phil and I somewhat endorse that plugin, like we do for your other plugins).

@herve91 :
- So please, clone this git repository under your own GitHub account (the easiest way being to use the GitHub fork feature which will create a copy for you). If you face any issues on this, I can help you.
- Furthermore, based on your message on the forum, my understanding is that this plugin is needed for your company only. In this context, it might be easier for your to create your own fork of Archi in which you directly patch parts of the code that you want to change. You'll them be able to locally distribute both your version of Archi and you plugin (which will then be lighter).

Regards,

JB

**Comment by herve91** [comment]
Hi JB,

Having several architects in several companies in the same holding, working on the same model is not unusual. Having a mecanism which avoid unwanted deletion or update on some Archimate concepts that are considered as "referential", avoiding dupplicates, managing concepts life cycle, comparison between branches (to be vs as is), ... are common need.

All my plugin started with my company needs and I shared them to the community as I believe that it may also fulfill some other companies need.

Until now, I centralized my plugins under archi-contribs as Phil and you recommanded, but in fact I don't mind where they are hosted. You may safely delete the repositories you wish under archi-contribs, I safely keep a copy of all my plugins sources.

Best regards
Hervé

**Comment by Phillipus** [comment]
@herve91 Please re-read https://github.com/archi-contribs/toolbox-plugin/issues/1#issuecomment-1674378888 for reasons why this plug-in is not endorsed and is, in fact, what is termed a "badly behaved plug-in". 

We will delete this repository within 24 hours.

**Comment by jbsarrodie** [comment]
>Having several architects in several companies in the same holding, working on the same model is not unusual. Having a mecanism which avoid unwanted deletion or update on some Archimate concepts that are considered as "referential", avoiding duplicates, managing concepts life cycle, comparison between branches (to be vs as is), ... are common need.

Yes, I have the same needs in my company and solve them using the model import feature (and coArchi, but your database plugin should do the job too): each "referential" is maintained in its own, dedicated, model (which only one or two architects can update), and then imported into "operational" models. Of course some people could then (by mistake) update/alter some of these info, but they are then rolled-back each time we re-import the "referential" models.

Locking concepts in the model is not necessarily needed, and, because of the way Archi is architectured at the moment, this is unfortunately not achievable through a plugin.

This might simply be the signal that Archi is no more the right tool for your company and you should think about purchasing a commercial tool which does support fine grained access control.

>All my plugin started with my company needs and I shared them to the community as I believe that it may also fulfill some other companies need.

And that's great, but you also have to face some responsabilities if your plugin behaves badly, which is the case of this specific one.

>Until now, I centralized my plugins under archi-contribs as Phil and you recommanded, but in fact I don't mind where they are hosted. You may safely delete the repositories you wish under archi-contribs, I safely keep a copy of all my plugins sources.

Ok, so we waiting some time to make sure everything is ok (just in case someone would want to clone the code in the meantime) and we'll delete this repository (for the sake of clarity: again, only this plugin is concerned, not your other plugins).
